### PR TITLE
List View: Add multi-select support for shift + Home and End keys

### PR DESCRIPTION
**Pull request by andrewserong**
## Description

Following on from https://github.com/WordPress/gutenberg/pull/38314, this PR adds in support for shift-selecting multiple blocks within the list view via the Home and End keys. The Tree Grid component already supports navigating to the beginning and end of the block list via these keys — this PR is a fairly simple update that rolls out adding in the `onFocusRow` callback for those key press events, too, and accepting those events as valid key presses for multiple block selection in `useBlockSelection`.

The end result is that it should now be a little easier to select from a particular block within the list view, to the end or beginning of a post.

## Testing Instructions

1. Add a range of blocks to a post. Open up the list view, and tab to focus on a block within the list.
2. Use Shift + Home or Shift + End keys to select a range of blocks.
3. Also use Shift + Up or Shift + Down keys to adjust the selection.
4. Check that this doesn't introduce any regressions.

Check that the added tests pass in CI (or you can use the following to run locally):

```
npm run test-unit packages/components/src/tree-grid/test/index.js
```

## Screenshots 

In the following screengrab, the Home and End keys allow selection to the beginning or end of the list while the shift key is selected. The up and down arrows then still enable altering that selection.

list-view-home-and-end-keys

## Types of changes

Enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

**Review by alexstine — COMMENTED** [review]
@andrewserong The change is almost perfect for me. The only issue is if I press Ctrl+End to select all blocks from my current position, it announces the number selected plus the last selected block. This really isn't helpful in this context. Can you ditch the last block name if selection is > than 1 based on the select Home/End keys event?

**Review by talldan — APPROVED** [review]
LGTM, thanks for working on another good improvement to List View.

From what I can tell, the issue @alexstine mentioned was solved by that last commit, but would be good to get a thumbs up from Alex before merging.

**Comment by andrewserong** [comment]
Thanks for the quick review @alexstine! I've pushed an update that skips announcing the deselected blocks when the new selection is generated via the Home/End keys and the selection is greater than 1. Is that what you had in mind? It's sounding better to me now using the macOS voice over feature, but I'm happy to keep tweaking if this change didn't quite address the issue 🙂

**Comment by alexstine** [comment]
@andrewserong Seems that the problem got worse. 👎 Here is what the screen reader announces.

> List View  toggle button  not pressed
pressed
Block library  region
Block navigation structure  table
Image  not selected  row 1  column 1
Image  link  Block 1 of 5, Level 1
5 blocks selected. 
Image (selected block)
Image (selected block)  not selected  row 5  column 1
Image (selected block)  link  Block 5 of 5, Level 1

Here all the extra selected blocks? Just the total would be enough.

1. I pulled latest changes.
2. Ran npm run build with no errors.
3. Opened my test site in private browsing.
4. Selected the List View button.
5. When focus was placed on the first block, I held Shift and pressed End once.

Hope this helps.

Thanks.

**Comment by ciampo** [comment]
I don't have much to add here, but just wanted to thank everyone involved (especially @alexstine — your feedback is always very useful, thank you!)

**Comment by andrewserong** [comment]
Thanks @alexstine, that feedback is very helpful! I've been testing with the voice over function in MacOS where the additional announcements of "Image (selected block)" aren't happening. I wonder if it's to do with that component, which is currently focused, being updated when the block selection changes. I'll see if I can install a Windows VM to try out NVDA or another screen reader to see if I can replicate what you've found in testing.

I'll follow up!

**Comment by andrewserong** [comment]
Alrighty, I'm wrapping up for the day, but I've set up a local environment running Windows 11 and NVDA and can reproduce the problem exactly as Alex described. As a side effect of this, I now have a much better testing environment! I'll do some exploration over the rest of the week to see what we can do to try to fix this up 👍

**Comment by andrewserong** [comment]
Alrighty, I think I've worked out _why_ there are additional announcements of the block that the user started on, and the destination block (I just don't have a fix for it yet 😅). In `block-select-button.js` the "(selected block)" text within the `VisuallyHidden` element updates when the selection changes, and given that the button is focused when that change happens, the currently focused button is announced again. The same appears to happen at the destination, the text has changed within the component so it is announced another time.

I'm not quite sure what the best workaround is for this, since we're relying on state changes for the screen reader to announce these components rather than a call to `speak`. I might not get a chance to continue investigating this week, but happy to continue investigating next week.

Thanks for rolling out the changes to `TreeGrid` and its tests from this PR in #39302 @alexstine!

**Comment by ramonjd** [comment]
I don't have a screen reader set up yet, but the interactions are working well for me. `SHIFT` + `UP/LEFT` or `DOWN/RIGHT` all select up and downwards from any position in a leaf node.

The only thing I'm not sure about – and it might be just because I'm using a mac keyboard – is that I can't trigger home or end (I assume it's meant to select the entire block?) 

The first few seconds of the screen case is me trying the [documented](https://support.apple.com/en-gb/guide/mac-help/cpmh0152/mac) `fn` + `LEFT/RIGHT` for `HOME/END`



 Any tips?

Non-blocking. It's a helpful interaction to have `SHIFT` alone.

File: packages/block-editor/src/components/list-view/use-block-selection.js
Comment (tellthemachines) [inline packages/block-editor/src/components/list-view/use-block-selection.js:129]: In testing on both Windows with NVDA+Firefox and macOS with VoiceOver+Firefox, this change doesn't seem to make any difference. The last selected block of the multi-selection is always announced, whether the multi-selection happens with Home/End keys or just Shift+Arrow Up/Down. And if for @alexstine the change seemed to make things worse, I think it's safe to remove it altogether.

**Review by tellthemachines — COMMENTED** [review]
Thanks for taking this on! The multi-selection behaviour with Shift+Home/End is working fine with the keyboard on all browsers except for Safari. (@ramonjd if you were testing on Safari, that wonky behaviour with Fn + Arrow Left/Right seems to be browser-specific; it's working fine on Firefox and Chrome)

The last selected item being announced by screen readers on multi-selection is also reproducible on trunk when selecting multiple items with Shift+Arrow keys, so I don't think it should be a blocker for this PR. We can investigate it separately.

File: packages/block-editor/src/components/list-view/use-block-selection.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/use-block-selection.js:129]: This change is good when selecting with Arrows but when using Home/End, it should only announce the number of selected/deselected. Why? Let's look at this practically.

- I press Shift+Down Arrow to select a block.
- I hear the total number of blocks selected plus the block I just selected.
- If I press End, it will select say 5 blocks but then I only hear the last block get announced.

See how this can get confusing?

When having these Home/End selections, it should just announce the total as in reality you selected multiple blocks at once vs. one block at a time.

Comment (tellthemachines) [reply packages/block-editor/src/components/list-view/use-block-selection.js:129]: I see what you mean. When pressing End, the last block in the selection gets focused; that must be why it's being announced. I wonder if we can instead keep focus on the first block in the selection - and if there won't be any negative side-effects to that 😅

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/use-block-selection.js:129]: Thanks for taking a closer look here! I was wondering if it's possible for us to conditionally add `aria-hidden` set to `true` in the block select button component for just the case of having switched focus while holding shift plus Home / End, to prevent the block from being announced. I might have a play.

Comment (talldan) [reply packages/block-editor/src/components/list-view/use-block-selection.js:129]: I think focus should always follow the selection. Otherwise when you press up or down again it becomes disorienting.

Looking at how Finder works, it's similar to what Alex describes (though you have to use shift+option+arrow key instead of home or end to select to the start or end). It'll only announce '6 rows selected' and it won't announce the file that is focused (even though focus does move).

Is there any way to suppress the innate announcement that screenreaders make when focus moves?

Comment (talldan) [reply packages/block-editor/src/components/list-view/use-block-selection.js:129]: I was also wondering if there's any advantage in using roles like `aria-multiselectable`:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-multiselectable

It didn't seem to do anything when I tested adding it, but maybe it needs to be used in combination with `aria-selected`.

The docs seem to imply that's ok with a `tree`, but don't say if that extends to a `treegrid`.

**Comment by andrewserong** [comment]
I tentatively think I might have come up with a fix for the announcement issue in https://github.com/WordPress/gutenberg/pull/39272/commits/2692f44b64f4fe9ff966feb4f9d1e0bf4390df29. Thanks again for the feedback and discussion, folks! I've rebased to resolve some conflicts with `trunk`.

In that change, I've added some state to the parent list view to keep track of whether or not we should prevent announcing the label of the block selection button. When we shift select with Home or End, we set `preventAnnouncement` to `true` and pass this all the way down to the block selection button. There, we pass this value to `aria-hidden` to prevent announcing the label for the button.

Then, when we move focus or select any other row, the `aria-hidden` flag is switched off. I've added a small delay, so that it isn't switched off prematurely when moving out of focusing on a particular button.

This appears to be testing pretty well for me so far in NVDA on Windows 11 running in a VM on a Mac. But I'd love a confidence check on the approach if you get a chance to take this for another spin @alexstine. No rush at all, though, as I'm finishing up for the week.

Also, very happy to throw out that approach if folks think there's a better way to do it, or if this approach feels too hacky!

**Comment by alexstine** [comment]
@andrewserong Seems to be working for me.

I did notice another bug though or maybe it is just something I missed the first time. There appears to be no way to deselect all blocks.
1. Open list view.
2. Select the first 2 blocks.
3. Try to deselect the first 2 blocks.
4. Because of the vertical navigation boundary, you cannot deselect the first block. At least this is what it seems to me.

The same happens with Home/End. You can select all blocks but you cannot remove the selection for all blocks.

Thoughts?

**Comment by alexstine** [comment]
Okay, writing a follow-up. On further review, this still appears to not be working with NVDA. The problem with a fairly hacky solution is screen readers may not honor it. Here is some output from the speech console.

```
List View  toggle button  not pressed
pressed
Block library  region
Block navigation structure  table
Paragraph  not selected  row 1  column 1
Paragraph  link  Block 1 of 9, Level 1
9 blocks selected. 
Column  not selected  row 13  column 1
Column  link  Block 2 of 2, Level 2
9 blocks selected.  
Paragraph  not selected  row 12  column 1
Paragraph  link  Block 1 of 1, Level 3
9 blocks selected. 
Column  not selected  row 13  column 1
Column  link  Block 2 of 2, Level 2
8 blocks deselected. 
selected  row 1  column 1
```

Essentially, I did the following.
1. Select the List View button.
2. Select all with Shift+Down Arrow.
3. Shift+Up Arrow to deselect the last block and then Shift+Down Arrow to select the last block again.
4. Shift+Home to deselect all blocks.

With this in mind, does the speech output make sense to you? I find it pretty confusing. 👎 I wish we could make this work like @talldan mentioned about the Mac Finder.

I know the re-usability aspect of this has got to be somewhat of a nightmare @andrewserong . Appreciate your work on researching solutions. We'll find one eventually.

Sorry it took me a while to get my thoughts cleared up. Sometimes I just need some time to really think about these more complex issues.

Thanks.

**Comment by tellthemachines** [comment]
Tested this again on Safari+VoiceOver (the home key combo is now working fine on Safari, not sure what happened earlier) and on Firefox+NVDA. In both the currently focused block is read out after "x blocks selected".

As Dan mentioned the `multiselectable` above, I decided to try it out in and put together #39600 as an experiment (I really just wanted a PR up so I could test it on Windows with Gutenberg.run 😅). It doesn't seem to change much in Safari+VO though. 

Not quite sure what else we can try here.

**Comment by andrewserong** [comment]
Thanks again for all the feedback and testing @tellthemachines, @alexstine, and @talldan!

I did some more investigation, thanks for the link to the alternate PR @tellthemachines, that sparked some ideas 😀

Just a heads-up, there's been a rebase recently, so you may need to remove your local copy of this branch before re-testing, in case that's caused any issues.

I think the root of our challenges here is that the set of roles that we're working with in the tree grid is similar to that of working within a grid of tabular data (e.g. the [row](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/row_role) role or the [gridcell](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/gridcell_role) role. Rather than wrestle too much with the browser here, I've had a go at reverting the `preventAnnouncement` approach, and instead see how far we can get by focusing on the row and cell behaviour of the tree grid.

In the latest update in https://github.com/WordPress/gutenberg/pull/39272/commits/159b2a2b6eacfbe04c0a6fea28f818b947f02a17, I've done a few things:

* Removed the `(selected block)` text at the button level, as this causes the block to be re-rendered when the text updates, which then triggers additional announcements when we select or deselect the block. Instead, like in @tellthemachines' PR, we move to using the `aria-selected` attribute on the cell and row. The missing piece appeared to be removing the dynamic changes and description at the button level.
* I've moved the description and labels up to the table cell, so that the focused block isn't announced twice. With the cell and row approach to the tree grid, the browser's default behaviour is to announce the label of the cell that is focused as well as the link contained within it. This appears to be the source of the multiple announcements of the focused block.
* So, with the label and description attached to the grid cell, I've then set the button itself to `aria-hidden` so that we only get a single announcement of each block.

This appears to be testing much better from my perspective using NVDA in Windows 11, where navigating between blocks now only reads that block out once. Shift-selecting all the blocks now no-longer reads out the block you started on a second time (without us having to add the hacky `preventAnnouncement` approach).

There is the remaining issue that we cannot appear to prevent the destination block from being announced after the selection. This appears to be browser default behaviour of navigating around cell-like elements, so it's a tricky one! I'm wondering if it's enough of an improvement that the block is only announced once instead of twice?

With the current state of the PR, the output from NVDA now sounds like the following in my testing:

```
Block navigation structure  table
Quote  not selected  Block 13 of 13, Level 1
Heading  not selected  Block 1 of 13, Level 1  row 1
Heading  not selected  Block 2 of 13, Level 1  row 2
12 blocks selected.  
Quote  selected  Block 13 of 13, Level 1  row 13
```

Please let me know if you think this direction seems viable — my hunch is that focusing on having the announcements at the cell or row level will help keep the behaviour contained and focused, but again, happy to throw any of this out if it doesn't seem suitable!

Thanks again for all the help and patience as we explore the different options 🙇

**Comment by alexstine** [comment]
@andrewserong It is better, there is no doubt about it. This is the problem I am seeing now. It might honestly be out of scope for this PR.

```
List View  toggle button  not pressed
pressed
Block library  region
Block navigation structure  table
Paragraph  not selected  Block 1 of 9, Level 1  row 1  column 1
9 blocks selected. 
Column  not selected  Block 2 of 2, Level 2  row 13  column 1
8 blocks deselected. 
Paragraph  selected  Block 1 of 9, Level 1  row 1  column 1
Paragraph  not selected  Block 2 of 9, Level 1  row 2  column 1
Paragraph  not selected  Block 3 of 9, Level 1  row 3  column 1
Paragraph  not selected  Block 4 of 9, Level 1  row 4  column 1
Paragraph  not selected  Block 5 of 9, Level 1  row 5  column 1
Paragraph  not selected  Block 6 of 9, Level 1  row 6  column 1
Paragraph  not selected  Block 7 of 9, Level 1  row 7  column 1
Group  not selected  expanded  Block 8 of 9, Level 1  row 8  column 1
Paragraph  not selected  Block 1 of 1, Level 2  row 9  column 1
Columns  not selected  expanded  Block 9 of 9, Level 1  row 10  column 1
Column  not selected  expanded  Block 1 of 2, Level 2  row 11  column 1
Paragraph  not selected  Block 1 of 1, Level 3  row 12  column 1
Column  not selected  Block 2 of 2, Level 2  row 13  column 1
```

> Column  not selected  Block 2 of 2, Level 2  row 13  column 1

In another PR, we've got to fix the label on this. This has confused me so bad, I don't even know what to make of that. We need to use simpler language somehow instead of expecting users to actually know what this means.

> 8 blocks deselected.

Why can I not deselect all blocks? Is it because the paragraph block is selected in the editor but not the selection as far as copy/paste? This confuses me why it says selecting 9 blocks but deselecting 8.

Thanks and great work on this!

**Comment by andrewserong** [comment]
Thanks for confirming, Alex! Since this approach seems to improve things, I'll proceed now with updating the e2e tests to get them passing (they're currently expecting the `selected block` string at the button link level).

> In another PR, we've got to fix the label on this.

Agreed. At least we've got some more information now on how to change things, so the hard part will be just deciding what we want it to say 🙂

> Why can I not deselect all blocks?

Oh, that's a good point, sorry I meant to follow-up on your comment there yesterday. I think the problem is that in comparable interfaces (such as a spreadsheet or file system) there is always at least one item currently selected, whereas in this interface there is such a thing as a focused but not selected item, which means we've got some more complexity to deal with. It's hard to know if when a user goes back to the original item with the shift key selected, whether or not they would like that item to remain selected. At the moment the logic assumes that the shift key means the original item is always going to be selected. Perhaps it'll be easier in a separate PR to look at the desired behaviour for de-selecting items, and explore which options might make sense? I'd be happy to look into that in a follow-up if it isn't a blocker for this PR.

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (tellthemachines) [inline packages/block-editor/src/components/list-view/block-select-button.js:84]: This may have been here for VoiceOver support, because now it's removed we don't get "expanded" or "collapsed" announced on focusing a collapsible item with Safari/VO (it works on Firefox, for some reason). TreeGrid semantics specify the `aria-expanded` should be on the `tr` or `td`, so it's correct to not have it here.

Not wanting to open a can of worms 😅 I think it should be fine to remove it. VO still announces "expanded" and "collapsed" when the status changes, so it's not terrible, and this change improves the experience for Windows/NVDA users.

File: packages/block-editor/src/components/list-view/use-block-selection.js
Comment (tellthemachines) [inline packages/block-editor/src/components/list-view/use-block-selection.js:128]: I'm wondering if we still need this here. Hypothetically, if we were to deselect one block at the same time as selecting 3 more, wouldn't we still want the deselected block to be announced?

**Comment by alexstine** [comment]
@andrewserong Follow-up PR is certainly okay. Super excited with the progress here.

Let's keep these improvements coming. 👍 Thanks!

**Review by tellthemachines — COMMENTED** [review]
Thanks for your continued work on this! The latest batch of changes is testing much better than previously; I just left a couple of comments below.

Regarding the inability to fully deselect list items, I think that should be addressed separately because it's an existing issue, not introduced by this PR.

@alexstine, is the `Block 2 of 2, Level 2 row 13 column 1` part too verbose? Perhaps we don't need to indicate whereabouts on the tree every time we move focus around. I think that would also be a good candidate for a follow-up though. It'd be good to get this in, as it's already better than what's currently on trunk!

**Comment by alexstine** [comment]
@tellthemachines I totally agree.

> @alexstine, is the Block 2 of 2, Level 2 row 13 column 1 part too verbose? Perhaps we don't need to indicate whereabouts on the tree every time we move focus around. I think that would also be a good candidate for a follow-up though. It'd be good to get this in, as it's already better than what's currently on trunk!

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/block-select-button.js:84]: @tellthemachines You are chasing after me with that comment. 👍 🙂 I added that to get around an ARIA bug which only happens when a link or button is in play. If link or button, aria-expanded has to be added directly. The `<tr>` or `<td>` wasn't enough. That's the context behind that one.

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (alexstine) [inline packages/block-editor/src/components/list-view/block-select-button.js:68]: @andrewserong What if you only added `aria-hidden="true"` during selection? The issue is, it no longer announces a link to screen readers so users may not understand this block title is clickable. Worth a test?

**Review by alexstine — COMMENTED** [review]
An interesting discovery I did not see since I was so busy checking selection.

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block-select-button.js:68]: From my explorations yesterday, in order to remove the duplicate announcements of the block, I think we need to permanently set `aria-hidden` to `true` at the button level. Otherwise the cell reads out the label from the child, and then the link is announced as well. If we'd like to better highlight that it's a link, I wonder if we can update the `label` that we're setting on the cell? At the moment I've just used the block title, but we could make this anything we like.

File: packages/block-editor/src/components/list-view/use-block-selection.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/use-block-selection.js:128]: I think in Alex's previous feedback, it sounded like it was adding too much to announce the deselected blocks at the same time. But I'm happy either way! @alexstine

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/block-select-button.js:68]: @andrewserong Can you add the word "link"?

Thanks.

File: packages/block-editor/src/components/list-view/use-block-selection.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/use-block-selection.js:128]: @andrewserong Could try again now that we've got a more solid solution. Single commit so could be easy revert if needed.

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/use-block-selection.js:128]: Sounds good, I'll have a play with removing it 👍

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/use-block-selection.js:128]: I've removed it in https://github.com/WordPress/gutenberg/pull/39272/commits/4dc20b9aa03ecdef9e638b58c3bbf4c14494335a, but like you mention, easy enough to add back in if we want to!

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block-select-button.js:84]: I've put this back in just in case, since it doesn't appear to affect whether or not the button is announced, as far as I can tell.

**Comment by andrewserong** [comment]
Alrighty, I've updated this PR with the following:

* Updated e2e tests to look for the element with `aria-selected` now that the selected text has been moved up to the cell.
* Added the word `link` to the label used at the cell level
* Removed the code to skip saying `deselected`

Please let me know if I've missed anything, or if it needs further tweaking!

File: packages/block-editor/src/components/list-view/use-block-selection.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/use-block-selection.js:128]: @andrewserong Probably best to add this back. The deselect announcement is a bit much I think considering it does announce what is currently selected at that point.

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/block-select-button.js:84]: @andrewserong No, it should not now that the link is hidden via aria-hidden. Up to you if you wish to keep aria-expanded on this now.

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block-select-button.js:84]: Ah, good point. For clarity sake, I'll remove it again.

File: packages/block-editor/src/components/list-view/use-block-selection.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/use-block-selection.js:128]: Okie-doke, I'll add it back in.

**Review by alexstine — APPROVED** [review]
Thanks for your work on this @andrewserong !

**Comment by andrewserong** [comment]
Thanks for reviewing again @alexstine! I've added back in the code to skip announcing the deselect, and removed the `isExpanded` at the button level.

I'll merge this in after lunch, once the tests pass!

**Comment by alexstine** [comment]
@andrewserong Sounds good. Then we can move on to more fun with this selection stuff. Maybe it would be a good idea to start a tracking issue if one doesn't already exist. Still a couple weird things to discuss. If you want to draft it, I can start adding some things to check unless you remember my spread out notes.

**Comment by andrewserong** [comment]
Sure thing, I'll open up a tracking issue a little later on today, and will ping you on it!

**Comment by andrewserong** [comment]
I've opened up [a follow-up issue](https://github.com/WordPress/gutenberg/issues/39671) to track some of the suggestions and ideas raised in the reviews here. Let me know, or add to that issue if I missed anything!

**Comment by mtias** [comment]
Thank you all for the great work here!

**Pull request merged**
